client/api_off_chain_auth.go

```diff
@@ -144,6 +189,130 @@ func (c *Client) RegisterEDDSAPublicKey(spAddress string, spEndpoint string) (st
+	userEddsaPublicKeyStr := hex.EncodeToString(userEddsaPublicKey)
+	log.Info().Msg("userEddsaPublicKeyStr is " + userEddsaPublicKeyStr)
+
+	IssueDate := time.Now().Format(time.RFC3339)
+	// ExpiryDate format := "2023-06-27T06:35:24Z"
+	ExpiryDate := time.Now().Add(time.Hour * 24).Format(time.RFC3339)
+
+	unSignedContent := fmt.Sprintf(unsignedContentTemplateV2, appDomain, c.defaultAccount.GetAddress().String(), userEddsaPublicKeyStr, appDomain, IssueDate, ExpiryDate)
+
+	unSignedContentHash := accounts.TextHash([]byte(unSignedContent))
+	sig, _ := c.defaultAccount.GetKeyManager().Sign(unSignedContentHash)
+	authString := fmt.Sprintf("%s,SignedMsg=%s,Signature=%s", httplib.Gnfd1EthPersonalSign, unSignedContent, hexutil.Encode(sig))
+	authString = strings.ReplaceAll(authString, "\n", "\\n")
+	headers := make(map[string]string)
+	headers["x-gnfd-app-domain"] = appDomain
+	headers["x-gnfd-app-reg-public-key"] = userEddsaPublicKeyStr
+	headers["X-Gnfd-Expiry-Timestamp"] = ExpiryDate
+	headers["authorization"] = authString
+	headers["origin"] = appDomain
+	headers["x-gnfd-user-address"] = c.defaultAccount.GetAddress().String()
+	jsonResult, error1 := httpPostWithHeader(spEndpoint+"/auth/update_key_v2", "{}", headers)
+
+	return jsonResult, error1
+}
+
+// ListUserPublicKeyV2 - List user public keys for off-chain-auth v2
+// This API will list user public keys for off-chain-auth v2. So that users/dapp can know what public keys have already been registered in a given sp.
+//
+// - spEndpoint: The sp endpoint where the list API will send the request.
+//
+// - domain: The domain that this list api will query for.
+//
+// - ret1: The public key list for the given sp/account/domain. The account will be the client's defaultAccount.
+//
+// - ret2: Return error when ListUserPublicKeyV2 runs into failure.
+func (c *Client) ListUserPublicKeyV2(spEndpoint string, domain string) ([]string, error) {
+	header := make(map[string]string)
+	header["X-Gnfd-User-Address"] = c.defaultAccount.GetAddress().String()
+	header["X-Gnfd-App-Domain"] = domain
+
+	response, err := httpGetWithHeader(spEndpoint+"/auth/keys_v2", header)
+	if err != nil {
+		return nil, err
+	}
+	listResp := ListUserPublicKeyV2Resp{}
+	// decode the xml content from response body
+	err = xml.NewDecoder(bytes.NewBufferString(response)).Decode(&listResp)
+	if err != nil {
+		return nil, err
+	}
+	return listResp.PublicKeys, nil
+}
+
+// DeleteUserPublicKeyV2 - Delete user public keys for off-chain-auth v2
+// This API will delete user public keys for off-chain-auth v2.
+//
+// - spEndpoint: The sp endpoint where the delete API will send the request.
+//
+// - domain: The domain that this api will delete for.
+//
+// - publicKeys: The public keys that the invoker intends to delete
+//
+// - ret1: Return deletion result.
+//
+// - ret2: Return error when DeleteUserPublicKeyV2 runs into failure.
+func (c *Client) DeleteUserPublicKeyV2(spEndpoint string, domain string, publicKeys []string) (bool, error) {
```

Comment:
the user's input can be validated before sending to SP

Reply:
Those input will be validated in SP side actually.